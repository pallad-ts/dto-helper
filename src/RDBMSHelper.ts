import {Knex} from 'knex';
import {Mapper} from '@pallad/mapper';
import {ID, create} from "@pallad/id";
import {fromNullable, Maybe} from "@sweet-monads/maybe";

export class RDBMSHelper<T, TID extends Knex.Value> {
	constructor(private knex: Knex, private options: RDBMSHelper.Options<T, TID>) {
		this.options = {
			timestamps: {
				createdAt: 'created_at',
				updatedAt: 'updated_at'
			},
			...this.options,
		};
	}

	createQueryBuilder() {
		return this.knex(this.options.table);
	}

	async findById(id: string): Promise<Maybe<T>> {
		return this.findOne({id});
	}

	async findOne(query: any): Promise<Maybe<T>> {
		return fromNullable(
			await this.createQueryBuilder()
				.where(query)
				.first()
				.then((result: any) => {
					// eslint-disable-next-line no-null/no-null
					if (result === null || result === undefined) {
						return undefined;
					}
					return this.options.mapper.mapToLight(result);
				})
		);
	}

	async find(query: any): Promise<T[]> {
		return this.createQueryBuilder()
			.where(query)
			.then((result: any[]) => {
				return this.options.mapper.arrayMapToLight(result);
			})
	}

	async findByQueryBuilder(func: RDBMSHelper.QueryBuilderModifier): Promise<T[]> {
		const qb = this.createQueryBuilder();
		func(qb);
		return this.options.mapper.arrayMapToLight(await qb);
	}

	async findOneByQueryBuilder(func: (builder: Knex.QueryBuilder<any, any>) => Knex.QueryBuilder<any, any>): Promise<Maybe<T>> {
		const qb = this.createQueryBuilder();
		func(qb);
		return fromNullable(
			await qb
				.first()
				.then((result: any) => {
					// eslint-disable-next-line no-null/no-null
					if (result === null || result === undefined) {
						return undefined;
					}
					return this.options.mapper.mapToLight(result);
				})
		);
	}

	private get mapper() {
		return this.options.mapper;
	}

	async create<TInput>(input: TInput) {
		const id = this.options.idGenerator();

		const data = this.mapper.mapPartialToDark({id, ...input} as any);
		const returning = [];
		if (this.options.timestamps && this.options.timestamps.createdAt) {
			returning.push(this.options.timestamps.createdAt);
		}
		if (this.options.timestamps && this.options.timestamps.updatedAt) {
			returning.push(this.options.timestamps.updatedAt);
		}

		const saveResult = await this.createQueryBuilder()
			.insert(data, returning);

		return this.mapper.mapToLight({
			...data,
			id,
			...(saveResult[0] || {})
		});
	}


	async update<TInput>(input: TInput, modifier: RDBMSHelper.QueryBuilderModifier) {
		const data = this.options.mapper.mapPartialToDark(input as any);
		if (this.options.timestamps && this.options.timestamps.updatedAt) {
			data[this.options.timestamps.updatedAt] = this.knex.fn.now(3);
		}
		const qb = this.createQueryBuilder()
			.update(data);

		modifier(qb);
		await qb;
	}

	updateById<TInput>(id: TID, input: TInput, modifier?: RDBMSHelper.QueryBuilderModifier) {
		return this.update(input, qb => {
			qb.where('id', id);
			modifier && modifier(qb);
		});
	}

	async delete(modifier: RDBMSHelper.QueryBuilderModifier) {
		const qb = this.createQueryBuilder()
			.delete();

		modifier(qb);

		await qb;
	}

	deleteById(id: TID, modifier?: RDBMSHelper.QueryBuilderModifier) {
		return this.delete(qb => {
			modifier && modifier(qb);
			qb.where('id', id)
		});
	}

	static createStandard<T>(knex: Knex, options: RDBMSHelper.StandardOptions<T>) {
		const timestamps: RDBMSHelper.Timestamps = {};
		if (options.withCreatedAt ?? true) {
			timestamps.createdAt = options.mapper.getDarkNameFromLightName('createdAt' as any) as string;
		}
		if (options.withUpdatedAt ?? true) {
			timestamps.updatedAt = options.mapper.getDarkNameFromLightName('updatedAt' as any) as string;
		}

		return new RDBMSHelper(knex, {
			table: options.table,
			mapper: options.mapper,
			idGenerator: create,
			timestamps
		});
	}
}

export namespace RDBMSHelper {
	export interface Options<T, TID> {
		table: string;
		mapper: Mapper<T>;
		idGenerator: () => TID,
		timestamps?: Timestamps;
	}

	export interface Timestamps {
		createdAt?: string,
		updatedAt?: string
	}

	export interface StandardOptions<T> extends Pick<Options<T, ID>, 'table' | 'mapper'> {
		withCreatedAt?: boolean;
		withUpdatedAt?: boolean;
	}

	export type QueryBuilderModifier = (knex: Knex.QueryBuilder) => void;
}
