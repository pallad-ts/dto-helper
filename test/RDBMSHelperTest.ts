import {connections} from "./connections";
import {RDBMSHelper} from "@src/RDBMSHelper";
import {Mapper} from '@pallad/mapper';
import {create} from '@pallad/id';

namespace DTO {
	export class Full {
		readonly id!: string;
		readonly field!: string;
		readonly createdAt!: Date;
		readonly updatedAt!: Date;

		constructor(data: Full) {
			Object.assign(this, data);
		}
	}

	export class OnlyCreatedAt {
		readonly id!: string;
		readonly field!: string;
		readonly createdAt!: Date;

		constructor(data: OnlyCreatedAt) {
			Object.assign(this, data);
		}
	}

	export class OnlyUpdatedAt {
		readonly id!: string;
		readonly field!: string;
		readonly updatedAt!: Date;

		constructor(data: OnlyCreatedAt) {
			Object.assign(this, data);
		}
	}
}


describe('RDBMSHelper', () => {
	const connection = connections.postgres;
	connection.setupBeforeAndAfterAll();

	connection.setupAfterEach({
		truncateTables: ['full', 'only_created_at', 'only_updated_at']
	});

	let helper_full: RDBMSHelper<DTO.Full, string>;
	let helper_onlyCreatedAt: RDBMSHelper<DTO.OnlyCreatedAt, string>;
	let helper_onlyUpdatedAt: RDBMSHelper<DTO.OnlyUpdatedAt, string>;

	beforeEach(() => {
		helper_full = new RDBMSHelper<DTO.Full, string>(connection.connection, {
			mapper: Mapper.create<DTO.Full>()
				.useLightFactory(x => new DTO.Full(x))
				.registerMapping('id')
				.registerMapping('field')
				.registerMapping('createdAt', 'created_at')
				.registerMapping('updatedAt', 'updated_at'),
			idGenerator: create,
			table: 'full',
		});

		helper_onlyCreatedAt = new RDBMSHelper<DTO.OnlyCreatedAt, string>(connection.connection, {
			mapper: Mapper.create<DTO.OnlyCreatedAt>()
				.useLightFactory(x => new DTO.OnlyCreatedAt(x))
				.registerMapping('id')
				.registerMapping('field')
				.registerMapping('createdAt', 'created_at'),
			idGenerator: create,
			table: 'only_created_at',
			timestamps: {
				createdAt: 'created_at'
			}
		});

		helper_onlyUpdatedAt = new RDBMSHelper<DTO.OnlyUpdatedAt, string>(connection.connection, {
			mapper: Mapper.create<DTO.OnlyUpdatedAt>()
				.useLightFactory(x => new DTO.OnlyUpdatedAt(x))
				.registerMapping('id')
				.registerMapping('field')
				.registerMapping('updatedAt', 'updated_at'),
			idGenerator: create,
			table: 'only_updated_at',
			timestamps: {
				updatedAt: 'updated_at'
			}
		});
	});

	describe('finding by id', () => {
		it('success', async () => {
			const record = await helper_full.create({
				field: 'field'
			});

			const found = await helper_full.findById(record.id);

			expect(found.value)
				.toEqual(record);
		});

		it('fail', async () => {
			const found = await helper_full.findById('anyid');
			expect(found.isNone())
				.toBeTruthy();
		});
	});

	describe('finding one', () => {
		it('success', async () => {
			await helper_full.create({
				field: 'field'
			});

			const record2 = await helper_full.create({
				field: 'field2'
			});

			const found = await helper_full.findOne({field: record2.field});

			expect(found.value)
				.toEqual(record2);
		});

		it('fail', async () => {
			const found = await helper_full.findOne({field: 'some'});
			expect(found.isNone())
				.toBeTruthy();
		});
	});

	describe('deleting', () => {
		it('removing existing', async () => {
			const record = await helper_full.create({
				field: 'field'
			});

			await helper_full.deleteById(record.id);

			const found = await helper_full.findById(record.id);
			expect(found.isNone())
				.toBeTruthy();
		});


		it('removing non-existing', async () => {
			const record = await helper_full.create({
				field: 'field'
			});

			await helper_full.deleteById(record.id);
			// call twice
			await helper_full.deleteById(record.id);

			const found = await helper_full.findById(record.id);
			expect(found.isNone())
				.toBeTruthy();
		});
	});

	describe('creating', () => {
		it('with all dates', async () => {
			const record = await helper_full.create({
				field: 'field'
			});

			expect(record)
				.toMatchSnapshot({
					id: expect.stringMatching(/^.{26}$/),
					createdAt: expect.any(Date),
					updatedAt: expect.any(Date)
				});
		});

		it('with create date only', async () => {
			const record = await helper_onlyCreatedAt.create({
				field: 'field'
			});

			expect(record)
				.toMatchSnapshot({
					id: expect.stringMatching(/^.{26}$/),
					createdAt: expect.any(Date)
				});
		});

		it('with update date only', async () => {
			const record = await helper_onlyUpdatedAt.create({
				field: 'field'
			});

			expect(record)
				.toMatchSnapshot({
					id: expect.stringMatching(/^.{26}$/),
					updatedAt: expect.any(Date)
				});
		});
	});

	describe('updating', () => {
		it('with all dates', async () => {
			const record = await helper_full.create({
				field: 'field'
			});

			await helper_full.updateById(record.id, {
				field: 'field2'
			});

			const found = await helper_full.findById(record.id);

			expect(found.value)
				.toHaveProperty('field', 'field2');
			expect(found.value!.updatedAt)
				.toBeAfter(record.updatedAt);
		});

		it('with created date only', async () => {
			const record = await helper_onlyCreatedAt.create({
				field: 'field'
			});

			await helper_onlyCreatedAt.updateById(record.id, {
				field: 'field2'
			});

			const found = await helper_onlyCreatedAt.findById(record.id);
			expect(found.value)
				.not
				.toHaveProperty('updatedAt');
		});

		it('with update date only', async () => {
			const record = await helper_onlyUpdatedAt.create({
				field: 'field'
			});

			await helper_onlyUpdatedAt.updateById(record.id, {
				field: 'field2'
			});

			const found = await helper_onlyUpdatedAt.findById(record.id);

			expect(found.value)
				.toHaveProperty('field', 'field2');

			if (found.isNone()) {
				throw new Error('Not `just`');
			}
			expect(found.value.updatedAt)
				.toBeAfter(record.updatedAt);
		});
	})
});
