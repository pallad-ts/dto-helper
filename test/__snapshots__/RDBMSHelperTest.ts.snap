// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RDBMSHelper creating with all dates 1`] = `
{
  "createdAt": Any<Date>,
  "field": "field",
  "id": StringMatching /\\^\\.\\{26\\}\\$/,
  "updatedAt": Any<Date>,
}
`;

exports[`RDBMSHelper creating with create date only 1`] = `
{
  "createdAt": Any<Date>,
  "field": "field",
  "id": StringMatching /\\^\\.\\{26\\}\\$/,
}
`;

exports[`RDBMSHelper creating with update date only 1`] = `
{
  "field": "field",
  "id": StringMatching /\\^\\.\\{26\\}\\$/,
  "updatedAt": Any<Date>,
}
`;
